project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java

```diff
@@ -0,0 +1,175 @@
+/********************************************************************************
+ * Copyright (c) 2022 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+package org.eclipse.glsp.example.javaemf.emf;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Optional;
+
+import com.google.inject.Inject;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.eclipse.emf.common.command.BasicCommandStack;
+import org.eclipse.emf.common.command.CommandStack;
+import org.eclipse.emf.common.util.EList;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.eclipse.emf.edit.domain.AdapterFactoryEditingDomain;
+import org.eclipse.emfcloud.modelserver.glsp.notation.Diagram;
+import org.eclipse.glsp.graph.GModelRoot;
+import org.eclipse.glsp.server.model.DefaultGModelState;
+import org.eclipse.glsp.server.model.GModelState;
+import org.eclipse.glsp.server.session.ClientSession;
+import org.eclipse.glsp.server.session.ClientSessionListener;
+import org.eclipse.glsp.server.session.ClientSessionManager;
+import org.eclipse.glsp.server.types.GLSPServerException;
+import org.eclipse.glsp.server.utils.ClientOptionsUtil;
+
+/**
+ * This state represents the status of the diagram based on the {@link GModelRoot}, contains the {@link EMFModelIndex},
+ * is able to execute commands via its {@link CommandStack} and is registered as {@link ClientSessionListener}
+ * to be able to reset the EMF resources on diagram close.
+ */
+public abstract class EMFGModelState extends DefaultGModelState implements ClientSessionListener {
+
+    private static Logger LOGGER = LogManager.getLogger(EMFGModelState.class.getSimpleName());
+
+    @Inject
+    protected ClientSessionManager clientSessionManager;
+
+    protected ResourceSet resourceSet;
+    protected AdapterFactoryEditingDomain editingDomain;
+
+    public abstract EObject getSemanticModel();
+
+    public abstract Diagram getNotationModel();
```

Comment:
I'd prefer if these two methods wouldn't be declared here, but rather be only implemented by subclasses.
Others who need these methods could get the subclass injected, if they need.
But this way this model state class is completely agnostic of how the source model(s) is/are structured.

Reply:
Agreed, moved that to the `TaskListGModelState` with the current commit.